components/authz-service/server/v2/rules_property_test.go

```diff
@@ -460,7 +463,108 @@ func TestDeleteRuleProperties(t *testing.T) {
 
 				return markedforDeletion && fullyDeleted
 			},
-			createProjectAndRuleGen,
+			createProjectAndRulesGen,
+		))
+
+	properties.TestingRun(t)
+}
+
+func TestListProjectProperties(t *testing.T) {
+	ctx := context.Background()
+	cl, testDB, _, _, seed := testhelpers.SetupProjectsAndRulesWithDB(t)
+	properties := getGopterParams(seed)
+
+	properties.Property("all staged rules are reported as staged",
+		prop.ForAll(
+			func(reqs projectAndRuleReq) bool {
+				defer testDB.Flush(t)
+
+				_, err := cl.CreateProject(ctx, &reqs.CreateProjectReq)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				reqs.rules, err = createRules(ctx, cl, reqs.rules)
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				resp, err := cl.ListProjects(ctx, &api.ListProjectsReq{})
+				if err != nil {
+					return reportErrorAndYieldFalse(t, err)
+				}
+
+				if 1 != len(resp.Projects) {
+					return false
+				}
+				respProject := resp.Projects[0]
+				t.Logf("===> %d rules", len(reqs.rules))
+				return respProject.Id == reqs.CreateProjectReq.Id &&
+					respProject.Name == reqs.CreateProjectReq.Name &&
+					respProject.Type.String() == "CUSTOM" &&
+					respProject.Status == storage.EditsPending.String()
+			},
+			createProjectAndRulesGen,
+		))
+
+	properties.TestingRun(t)
+}
+
+func TestListProjectPropertiesMixingStagedAndApplied(t *testing.T) {
+	ctx := context.Background()
+	cl, testDB, _, _, seed := testhelpers.SetupProjectsAndRulesWithDB(t)
+	properties := getGopterParams(seed)
+
+	properties.Property("projects with staged, applied, or no rules are reported",
+		prop.ForAll(
+			func(reqs []projectAndRuleReq) bool {
```

Comment:
How to read a property-based test. Step 1: look for the `func` line like this. Gopter feeds a set of test cases, one at a time, in the shape of the argument to this function. The shape was designed by me to specifically to test `ListProjects`. One test case, then, is a list of objects, each of which contains a single project and some rules (or no rules) to go with it.